crates/fuel-core/src/database/balances.rs

```diff
@@ -48,37 +59,91 @@ impl StorageMutate<ContractsAssets> for Database {
         key: &<ContractsAssets as Mappable>::Key,
         value: &<ContractsAssets as Mappable>::Value,
     ) -> Result<Option<<ContractsAssets as Mappable>::OwnedValue>, Self::Error> {
-        Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
-            .map_err(Into::into)
+        let prev = Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
+            .map_err(Into::into);
+
+        // Get latest metadata entry for this contract id
+        let prev_metadata = self
+            .storage::<ContractsAssetsMerkleMetadata>()
+            .get(key.contract_id())?
+            .unwrap_or_default();
+
+        let root = prev_metadata.root;
+        let mut tree: MerkleTree<ContractsAssetsMerkleData, _> = {
+            let storage = self.borrow_mut();
+            if root == [0; 32] {
+                // The tree is empty
+                MerkleTree::new(storage)
+            } else {
+                // Load the tree saved in metadata
+                MerkleTree::load(storage, &root)
+                    .map_err(|err| StorageError::Other(err.into()))?
+            }
+        };
+
+        // Update the contact's key-value dataset. The key is the asset id and the
+        // value the Word
+        tree.update(key.asset_id().deref(), value.to_be_bytes().as_slice())
+            .map_err(|err| StorageError::Other(err.into()))?;
+
+        // Generate new metadata for the updated tree
+        let root = tree.root();
+        let metadata = SparseMerkleMetadata { root };
+        self.storage::<ContractsAssetsMerkleMetadata>()
+            .insert(key.contract_id(), &metadata)?;
+
+        prev
     }
 
     fn remove(
         &mut self,
         key: &<ContractsAssets as Mappable>::Key,
     ) -> Result<Option<<ContractsAssets as Mappable>::OwnedValue>, Self::Error> {
-        Database::remove(self, key.as_ref(), Column::ContractsAssets).map_err(Into::into)
+        let prev = Database::remove(self, key.as_ref(), Column::ContractsAssets)
+            .map_err(Into::into);
+
+        // Get latest metadata entry for this contract id
+        let prev_metadata = self
+            .storage::<ContractsAssetsMerkleMetadata>()
+            .get(key.contract_id())?
+            .unwrap_or_default();
+
+        let root = prev_metadata.root;
+        if root == [0; 32] {
```

Comment:
It sounds like you're suggesting that we check the presence of the returned option, rather than the root. That could work, but there are two concerns that come to mind:
1. When we delete the key from the loaded Merkle tree below, we would have to check the root to see if it is empty. If it is empty, we must delete the Merkle metadata record for this contract. Otherwise, this check returns `Some(...)` rather than `None` in this case.
2. Comparing the root against the empty root is already specification compliant. According to the spec, that is a how we know that the Merkle tree is empty. Checking the DB return value will work too, but I don't know if it's actually simpler in this case.